ghga_validator/plugins/ref_validation.py

```diff
@@ -0,0 +1,249 @@
+                if range_class != class_name:
+                    continue
+                if isinstance(value, list):
+                    id_list = [x[id_slot_name] for x in value]
+                else:
+                    id_list = [value[id_slot_name]]
+        return id_list
+
+    def get_class_def(self, class_name) -> ClassDefinition:
+        """
+        Get class definition.
+
+        Args:
+            class_name: class name
+
+        Returns:
+            ClassDefinition: class definition
+
+        """
+        formatted_class_name = camelcase_to_sentencecase(class_name)
+        return self.schemaview.get_class(formatted_class_name)
+
+    def get_slot_def(self, class_name: str, slot_name: str) -> SlotDefinition:
+        """
+        Get slot definition.
+
+        Args:
+            slot_name: slot name
+
+        Returns:
+            SlotDefinition: class definition
+
+        """
+        formatted_slot_name = snakecase_to_sentencecase(slot_name)
+        class_def = self.get_class_def(class_name)
+        slot_usage = class_def.slot_usage
+        if formatted_slot_name in slot_usage:
+            return slot_usage[formatted_slot_name]
+        return self.schemaview.get_slot(formatted_slot_name)
+
+    def find_missing_refs(
+        self, ref_value: Union[List[str], str], id_list: List
+    ) -> List:
+        """
+        Search for missing references
+
+        Returns:
+            List: List of missing references
+        """
+        if not isinstance(ref_value, list):
+            value_to_check = [ref_value]
+        else:
+            value_to_check = ref_value
+        non_match = []
+        if not all(x in id_list for x in value_to_check):
+            non_match = [x for x in value_to_check if x not in id_list]
+        return non_match
+
+    def non_match_as_string(self, non_match: List[str]) -> str:
+        """
+        Generate a validation error message for missing references
+
+        Returns:
+            str: The validation error message
+        """
+        if len(non_match) > 1:
+            non_match_as_string = (
+                "'" + "', '".join(str(x) for x in non_match) + "'" + " were unexpected"
+            )
+        else:
+            non_match_as_string = "'" + non_match[0] + "'" + " was unexpected"
+        return non_match_as_string
+
+    def get_range_class(self, slot_def: SlotDefinition) -> Optional[str]:
+        """Return the range class for a slot
+
+        Args:
+            slot_def (SlotDefinition): Slot Definition
+
+        Returns:
+            Optional[str]: Range class for a slot
+        """
+        all_classes_name = self.schemaview.all_classes().keys()
+        if slot_def:
+            range_class = slot_def.range
+            if range_class in all_classes_name:
+                return range_class
+        return None
+
+    def to_list(self, value: Any) -> List[Dict]:
+        """
+        If an value is not an instance of List of objects, transform to it
```

Comment:
This sounds generic, but there is additional logic in this function, i.e. it expects that the first list element is a dictionary?

Reply:
It can be list, but also a single value - I replaced it with the Union to make it more clear